[pcf8574] Add interrupt pin support

gaaf opened the pull request:
# What does this implement/fix?

Add interrupt pin support to pcf8574 devices and only read from i2c when necessary. Disable the loop when there's nothing to read.

The first few patches prepare the code to properly support the interrupt pin as all unused pins must be set to outputs to prevent (millions of) spurious interrupts.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5165

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

All commits up to the loop disabling have been in production use for almost 2 years.

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
pcf8574:
  - id: 'pcf8574_hub'
    address: 0x21
    pcf8575: false
    interrupt_pin:
      number: 2
      mode: INPUT_PULLUP
      inverted: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

gaaf committed:
pcf8574: Only read from the device once every loop

Each input pin is read in every loop. For every pin, the entire byte/word
was read. With this change the i2c bus is read only once per loop (and only
if there are input pins). The individual pins read from the cached value.

This has several advantages;
1) Most important: all pins now have a coherent value during each loop
   invocation. If pins are related to each other, for example when reading
   a rotary encoder or a (binary encoded) numerical value from multiple
   pins. The single read will guarantee a coherent value between all pins
   on each loop.
2) Less contention on the i2c bus
3) Faster loop

gaaf committed:
pcf8574: Invert mode_mask

Make logic of setting output bits easier

gaaf committed:
pcf8574: Add interrupt pin handling

Only read from the device if the interrupt has triggered.

Try to read asap after an interrupt triggered.

gaaf committed:
pcf8574: Set unused port to outputs

If the unused floating ports are left as inputs, they have no defined
state and may trigger a huge amount of state changes. This causes
increased power usage and spurious interrupts on the INT port.

The TI datasheet mentions in §9.2.D:
"P6 and P7 are not used and must be configured as outputs."

So all unused port must be set to 0. There is one slight complication: if
we start with all ports as outputs, they might flap during setup. To prevent
that, all ports start out as inputs during setup(). The final mode_mask is
only set and written to the device after all setup completed, ie during the
first loop().

gaaf committed:
pcf8574: Rename input/output_mask to input/output_state

More representative of its function.

gaaf committed:
pcf8574: Rewite deprecated changes++ as changes = changes + 1

https://blog.feabhas.com/2021/05/modern-embedded-c-deprecation-of-volatile/

gaaf committed:
pcf8574; Set warning message on i2c error

gaaf committed:
pcf8574: Disable loop when nothing to read

1) When there are no inputs configured, loop has nothing to do
2) When interrupts are enabled, there's nothing to read until an interrupt
   fires.

gaaf committed:
pcf8574: Add whitespace for readability

DjordjeMandic commented on esphome/components/pcf8574/pcf8574.cpp:
```diff
@@ -16,86 +27,149 @@ void PCF8574Component::setup() {
+      return;
+    } else {
+      this->enable_loop();
+    }
```
```suggestion
    }
    this->enable_loop();
```

gaaf replied:
This code actually makes no sense at all as it can only be reached if loop is already enabled. I moved the enable_loop().

gaaf committed:
pcf8574: Move enable_loop() out of loop()

It makes no sense in loop() as the loop must already be enabled to reach
the code. Move it to where a pin mode can actually change during runtime.

gaaf commented:
I just remembered why I didn't submit this back in 2023: Some configurations might need to read pin changes during the loop and this PR forces only a single read per loop invocation.

I retract the PR for now. I'll resubmit if I find a clean way to support both use cases.
